### PR TITLE
[RW-2517][RISK=NO]Create Account page

### PR DESCRIPTION
**Pull request by NehaBroad**
This is First step of new create account. The next step will be part of story RW-3284 | Create Account - User survey pagewhich will have the submit/save account information functionality, Because of this reason i have set the environment variable 'enableAccountPages'  set to false even on local/test
Screen Shot 2019-09-05 at 5 59 52 PM
Screen Shot 2019-09-05 at 6 00 00 PM
Screen Shot 2019-09-05 at 6 00 07 PM

### CODE REVIEW

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 1]
```diff
@@ -0,0 +1,665 @@
...
+
+// The values will change once we have ENUM as part of DB work as part of create account 2
+const Options = {
+  roles: [
+    {label: `Undergraduate (Bachelor level) student`, value: 'bachelor'},
+    {label: `Graduate trainee (Current student in a Masters, PhD, or Medical school training
+        program)`, value: 'trainee'},
+    {label: `Research fellow (a post-doctoral fellow or medical resident in training)`,
+      value: 'research'},
+    {label: `Early career tenure-track researcher`, value: 'earylCarrer'},
```
Comment (blrubenstein) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: Nit: Spelling on value.

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 2]
```diff
@@ -0,0 +1,665 @@
...
+    {label: `Non tenure-track researcher`, value: 'nontenure'},
+    {label: `Mid-career tenured researcher`, value: 'midCarrer'},
```
Comment (blrubenstein) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: Nit: Spelling (should be `midCareer`), also line below

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+import * as fp from 'lodash/fp';
+import {Dropdown} from 'primereact/dropdown';
+import * as React from 'react';
+import * as validate from 'validate.js';
+
+function isBlank(s: string) {
+  return (!s || /^\s*$/.test(s));
+}
+
+// The values will change once we have ENUM as part of DB work as part of create account 2
```
Comment (blrubenstein) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: :+1: to this

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 4]
```diff
@@ -0,0 +1,665 @@
...
+  ]
+};
+
+// The following will be part of Profile object once create account survey is in
+export interface Address {
+  streetAddress1: string;
+  streetAddress2: string;
+  city: string;
+  state: string;
+  zipcode: string;
```
Comment (blrubenstein) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: Should the zipcode be a number? Functionally it seems like number might be right.

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 5]
```diff
@@ -0,0 +1,665 @@
...
+export interface AccountCreationProps {
+  invitationKey: string;
+  setProfile: Function;
+}
+
+export interface AccountCreationState {
+  creatingAccount: boolean;
+  errors: any;
+  invalidEmail: boolean;
+  rolesOptions: any;
```
Comment (blrubenstein) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: I think this could be more descriptive than `any`. Like maybe: `Option[]`? (and then create an Option interface?)

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 6]
```diff
@@ -0,0 +1,665 @@
...
+      .then((savedProfile) => {
+        this.setState({profile: savedProfile, creatingAccount: false});
+        setProfile(savedProfile); })
+      .catch(error => {
+        console.log(error);
+        this.setState({creatingAccount: false});
+      });
+  }
+
+  usernameValid(): boolean {
```
Comment (blrubenstein) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: Could this be a `get` method?

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+    if (isBlank(this.state.profile.username) || this.state.usernameCheckInProgress) {
+      return undefined;
+    }
+    return !this.isUsernameValidationError();
+  }
+
+  isUsernameValidationError(): boolean {
```
Comment (blrubenstein) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: Could this be a `get` method?

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 8]
```diff
@@ -0,0 +1,665 @@
...
+    return (this.state.usernameConflictError || this.usernameInvalidError());
+  }
+
+  usernameInvalidError(): boolean {
```
Comment (blrubenstein) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: Could this be a `get` method?

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 9]
```diff
@@ -0,0 +1,665 @@
...
+    } else if (affiliation === 'freeText') {
+      this.setState({showAffiliationOther: true});
+      return;
+    }
+    this.selectAffiliationRoles(this.state.affiliationRole);
+  }
+
+  selectAffiliationRoles(role) {
+    this.setState({affiliationRole: role});
+    if (role === 'freeText') {
```
Comment (blrubenstein) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: Opt Nit: Could you put this above and have it only issue one call to setState?

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 10]
```diff
@@ -0,0 +1,665 @@
...
+      [profile.givenName, profile.familyName, profile.username, profile.contactEmail,
+        profile.currentPosition, profile.organization, profile.areaOfResearch];
+    if (requiredFields.some(isBlank)) {
+      this.setState({showAllFieldsRequiredError: true});
+      return;
+    }
+  }
+
+  validateAccountCreation() {
+    const {profile: {
```
Comment (blrubenstein) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: Nit: For spacing, do you mind putting `profile` on a newline? This is sorta weirdly spaced with everything else in the object.

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 11]
```diff
@@ -0,0 +1,665 @@
...
+      givenName: {
+        presence: {allowEmpty: false}
+      },
+      familyName: {
+        presence: {allowEmpty: false}
+      },
+      contactEmail: {
+        presence: {allowEmpty: false},
+        format: {
+          pattern: /^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/,
```
Comment (blrubenstein) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: Could you use the same regex pattern on line 190 for email validation?

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 12]
```diff
@@ -0,0 +1,665 @@
...
+            {summarizeErrors(this.state.errors)}
+          </ValidationError>
+        </div>}
+        <Section header='Create an All of Us username'>
+          <div>
+            <TextInput id='username' name='username' placeholder='New Username'
+                       value={username}
+                       onChange={v => this.usernameChanged(v)}
+                       invalid={this.state.usernameConflictError || this.usernameInvalidError()}
+                       style={{width: '12rem', marginRight: '0.5rem'}}/>
```
Comment (blrubenstein) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: I'm seeing `width: '12rem'` repeated a bunch. Can we parameterize that?

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 13]
```diff
@@ -0,0 +1,665 @@
...
+            </div>
+            <TooltipTrigger content={<div>Usernames can contain only letters (a-z),
+              numbers (0-9), dashes (-), underscores (_), apostrophes ('), and periods (.)
+              (minimum of 3 characters and maximum of 64 characters).<br/>Usernames cannot
+              begin or end with a period (.) and may not contain more than one period (.) in a row.
+            </div>}>
+              <InfoIcon style={{'height': '22px', 'paddingLeft': '2px'}}/>
+            </TooltipTrigger>
+            {this.state.usernameConflictError &&
+            <div style={{height: '1.5rem'}}>
```
Comment (blrubenstein) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: I'm seeing this height repeated for most errors. Can we parameterize that?

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 14]
```diff
@@ -0,0 +1,665 @@
...
+            </Error></div>}
+          </div>
+        </Section>
+        <Section header='About you'>
+          <div style={{display: 'flex', flexDirection: 'column'}}>
+            <div style={{paddingBottom: '0.5rem'}}>
+              <TextInput id='givenName' name='givenName' autoFocus
+                         placeholder='First Name'
+                         value={givenName}
+                         invalid={givenName.length > 80}
```
Comment (blrubenstein) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: Can we parameterize 80 as something like `nameLengthLimit`?

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 15]
```diff
@@ -0,0 +1,665 @@
...
+                         style={{width: '12rem', marginRight: '2rem'}}
+                         onChange={v => this.updateProfile('givenName', v)}/>
+              {givenName.length > 80 &&
+              <ErrorMessage id='givenNameError'>
+                First Name must be 80 characters or less.
```
Comment (blrubenstein) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: If we parameterize the name length limit, these error messages can also read text from that.

File: ui/src/app/pages/login/account-creation/account-creation.tsx (hunk 11 above)
Comment (NehaBroad) [reply ui/src/app/pages/login/account-creation/account-creation.tsx]: Its the same regex, the other is going to be deleted once we have the enable flag set to true

File: ui/src/app/pages/login/account-creation/account-creation.tsx (hunk 4 above)
Comment (NehaBroad) [reply ui/src/app/pages/login/account-creation/account-creation.tsx]: For zipcode i was planning to use this as string so as to not loose the leading zeros

File: ui/src/app/pages/login/account-creation/account-creation.tsx (hunk 15 above)
Comment (blrubenstein) [reply ui/src/app/pages/login/account-creation/account-creation.tsx]: I think you could change 80 to `{{nameLength}}` (and below as well)

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 16]
```diff
@@ -439,7 +439,7 @@ export class AccountCreation extends React.Component<AccountCreationProps, Accou
                          onChange={v => this.updateProfile('givenName', v)}/>
               {givenName.length > nameLength &&
               <ErrorMessage id='givenNameError'>
-                First Name must be 80 characters or less.
+                First Name must be {nameLength} characters or less.
```
Comment (jaycarlton) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: nit: this should be fewer rather than less since it's countable in a reasonable period of time.

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 17]
```diff
@@ -0,0 +1,668 @@
...
+
+import colors from 'app/styles/colors';
+import {summarizeErrors} from 'app/utils/index';
+import {environment} from 'environments/environment';
+import * as fp from 'lodash/fp';
+import {Dropdown} from 'primereact/dropdown';
+import * as React from 'react';
+import * as validate from 'validate.js';
+
+function isBlank(s: string) {
```
Comment (jaycarlton) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: is there a builtin or library function for this? If we need to roll our own, then we need to have unit tests I think.

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 18]
```diff
@@ -0,0 +1,668 @@
...
+  return (!s || /^\s*$/.test(s));
+}
+
+// The values will change once we have ENUM as part of DB work as part of create account 2
+const Options = {
+  roles: [
```
Comment (jaycarlton) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: Can we externalize this into a JSON file for roleConfig or something?

File: ui/src/app/pages/login/account-creation/account-creation.tsx (hunk 4 above)
Comment (jaycarlton) [reply ui/src/app/pages/login/account-creation/account-creation.tsx]: nit: if this is for international addresses as well, we need to be careful. One (minor) point is that I think the US is the only one that calls it a zip code. The bigger problem is that we need a corpus of test cases to make sure we're representing them fully. Again, it seems like there would be an off-the-shelf component for working with addresses.

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 19]
```diff
@@ -0,0 +1,668 @@
...
+      affiliationOther: ''
+    };
+  }
+
+  // This method will be modified once the story  below will be done
+  // https://precisionmedicineinitiative.atlassian.net/browse/RW-3284
+  createAccount(): void {
+    const {invitationKey, setProfile} = this.props;
+    const profile = this.state.profile;
+    const emailValidRegex = new RegExp(/^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/);
```
Comment (jaycarlton) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: Please use a library to validate email addresses. There are tons of edge cases. For example, it looks like this won't support the plus character, which gmail uses for lightweight aliases.

We need lots of unit tests too.

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 20]
```diff
@@ -0,0 +1,668 @@
...
+    this.setState({showAllFieldsRequiredError: false});
+    this.setState({invalidEmail: false});
+    const requiredFields =
+      [profile.givenName, profile.familyName, profile.username, profile.contactEmail,
```
Comment (jaycarlton) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: Not everybody has two names. There's a classic post about this at https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/.

Likewise, for citizen scientists, they might not have a current position, org, or research area. [this may be a future issue].

Sorry with hitting you with product concerns in a PR.

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 21]
```diff
@@ -0,0 +1,668 @@
...
+        username: '',
+        dataAccessLevel: DataAccessLevel.Unregistered,
+        givenName: '',
+        familyName: '',
+        contactEmail: '',
+        currentPosition: '',
+        organization: '',
+        areaOfResearch: '',
+      },
+      usernameCheckInProgress: false,
```
Comment (jaycarlton) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: I believe this state could be tracked in a more object-oriented way, for example by having a little component for each piece of data we're taking in and having it report errors (and be able to define what its own error state looks like). That way we could make them smarter later (e.g. allowing incremental feedback while you're typing, for characters remaining, forbidden characters, etc.). The way it is now, there's no real separation among all the input widgets' state, and the maintainer has to keep track fo whether each one is checked/flagged. 

This is just a quick look though; you could have reasons it needs to be centralize like this too.

**Comment by jaycarlton** [comment]
@NehaBroad I don't really understand the state we're in now or why we can't use a feature branch or something, but go ahead if these files aren't new. I need to pay more attention.

**Comment by blrubenstein** [comment]
@jaycarlton we have tended to avoid feature branches, since they can become stale quickly, and often lead to several awful rebases, in favor of smaller PRs behind flags. It does make unit testing hard though.

**Comment by jaycarlton** [comment]
@blrubenstein OK. I don't know much about how the feature flags work, but at other places where we've used them, unit test coverage was one of the things we relied on before turning them on. Is there a way to mock the feature flag for testing purposes, or to un-flag a feature to only the test account? Sorry for thread-jacking.

**Comment by gjuggler** [comment]
> @jaycarlton we have tended to avoid feature branches, since they can become stale quickly, and often lead to several awful rebases, in favor of smaller PRs behind flags. It does make unit testing hard though.

If we can't write unit tests for code that's behind a feature flag, this is an important bug / friction point in our infrastructure. I had a review yesterday with @als364 where we ran into the same thing. I filed https://precisionmedicineinitiative.atlassian.net/browse/RW-3457 to track this as tech debt.
